File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque
input = sys.stdin.readline

dx = [-2, -2, -1, -1, 1, 1, 2, 2]
dy = [1, -1, 2, -2, 2, -2, 1, -1]

def bfs(x, y):
    queue = deque([(x, y)])
    chess[y][x] = 0
    while queue:
        x, y = queue.popleft()
        if x == target_x and y == target_y:
            return chess[y][x]
        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < size and 0 <= ny < size and chess[ny][nx] == 0:
                chess[ny][nx] = chess[y][x] + 1
                queue.append((nx, ny))
        
                
case = int(input())
for _ in range(case):
    size = int(input())
    chess = [[0 for _ in range(size)] for _ in range(size)]
    now_x, now_y = map(int, input().split())
    target_x, target_y = map(int, input().split())
    print(bfs(now_x, now_y))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cc
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

const int MAX = 300;
int dx[8] = {-2, -2, -1, -1, 1, 1, 2, 2};
int dy[8] = {1, -1, 2, -2, 2, -2, 1, -1};

int bfs(int size, int now_x, int now_y, int target_x, int target_y) {
    vector<vector<int>> chess(size, vector<int>(size, 0));
    queue<pair<int, int>> q;

    q.push({now_x, now_y});
    chess[now_y][now_x] = 0;

    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        if (x == target_x && y == target_y)
            return chess[y][x];

        for (int i = 0; i < 8; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx >= 0 && nx < size && ny >= 0 && ny < size && chess[ny][nx] == 0) {
                chess[ny][nx] = chess[y][x] + 1;
                q.push({nx, ny});
            }
        }
    }
    return -1;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;

    while (t--) {
        int size, now_x, now_y, target_x, target_y;
        cin >> size;
        cin >> now_x >> now_y;
        cin >> target_x >> target_y;

        cout << bfs(size, now_x, now_y, target_x, target_y) << "\n";
    }

    return 0;
}